old-archive/interfaces/ebuttons-1.0/ebuttons.1
.TH EBUTTONS 1 "18 November 1992"
.SH NAME
ebuttons \- popup buttons for giving commands to an emacs session.
.SH SYNOPSIS
.B M-x ebuttons
inside emacs, or
.B (ebuttons) 
in an emacs startup file.
.SH DESCRIPTION
.B Ebuttons
provides an X interface to issue commands to an emacs
session.  It allows you to specify (in an X resource file) labels for
a set of buttons and a command for each that will be executed when the
corresponding button is clicked on with the mouse. For instance you
can define buttons to compile, to find the next error, to save
buffers, to move to the top/bottom of the buffer, to exit emacs etc.

The idea is that this should be a thin vertical window that sits just
to the side of an emacs session and allows for fast execution of some
things. It is not intended to replace the keyboard! It just gives
another option - if your hand is already on the mouse, use 
.BR ebuttons .
If not, use the keyboard. Or as you like it. The window can also be
made horizontal (see below) but then not as many buttons will fit on
the screen.
.SH STARTUP
The simplest method is to automatically start 
.B ebuttons
when you enter emacs. This can be done by adding the following to your .emacs file.
.sp
    (if (equal window-system 'x)
.br
        (progn
.br
	      (load "ebuttons")
.br
	      (ebuttons)))
.SH "KEY BINDINGS"
It is useful to bind (ebuttons) to some key inside emacs. This function 
starts the program if it is not already running 
and toggles its visibility otherwise. The simplest way to do this is
to put something like
.sp
    (define-key global-map "\^Xt" 'ebuttons)
.sp
in your .emacs file. This binds the command to the keystroke control-X t.
.SH RESOURCES
Your resource file should have pairs of lines like
.sp
    ebuttons.labelX:         label
.br
    ebuttons.commandX:       (emacs-command)
.sp
where X is any number from 0 to the number of possible buttons minus
one. The upper limit is set in the Makefile and is 50 in the
distribution. Label numbers may be omitted (i.e. you can define labels
1, 2 and 5, omitting 3 and 4). If you only define 3 buttons, only 3
will appear. Actually this is not entirely true, an extra button (the
last button in the window) is added. It causes the 
.B ebuttons 
program to
exit and is (usually) labelled Quit.  You can change the name that
appears on this button (if you wish) with:
.sp
    ebuttons*Quit.label: your-label
.sp
in your resource file. Other resources that will have useful effects
are
.sp
    ebuttons*Command.borderWidth
    ebuttons*Box.orientation
    ebuttons*Box.hSpace
    ebuttons*Box.vSpace
    ebuttons.geometry
    ebuttons*background
    ebuttons*font
.sp
The orientation resource can be set to "horizontal" or "vertical" (the default).
Look in the Resources file in the distribution for an example of settings for these
resources. My preference is to have the 
.B ebuttons 
window take up as little space as
possible, so I use a small font and leave no vertical or horizontal
space between the command buttons.
.SH "SEE ALSO"
The README in the distribution and
.B Taglist 
by Brad Mears (bmears@deltahp.jsc.nasa.gov) from which this program
was distilled.
.SH AUTHOR
Terry Jones (terry@santafe.edu)
.br
Santa Fe Institute
.br
November 18, 1992.


NEWS
Ohio State Emacs Lisp Archive NEWS

* Fri Nov  2 17:55:35 2001

I forgot to update this file after updating the web page; the OSU
Emacs Lisp archive is closed while a merge with Lispmeralda
(www.lispmeralda.org) takes place.  We hope to accomplish this soon.
Thanks for your patience!

* Wed Jun 28 18:17:05 EDT 2000

Good news!  The Ohio State Emacs Lisp Archive is now being actively
maintained again.  We are still trying to catch up on organizing the
entries.  Our current project is to create a common archive entry
format.  However, submissions are being taken, currently by hand.
Please see the file SUBMISSIONS.

Send comments to the Emacs Lisp Archive maintainers:
archive@cis.ohio-state.edu

Colin Walters <walters@cis.ohio-state.edu>

Local Variables:
mode:text
End:


old-archive/interfaces/ebuttons-1.0/ebuttons.c
#ifndef lint
static char *rcsid = "$Header: /tmp_mnt/vida/disks/disk5/Users/terry/s/ebuttons/RCS/ebuttons.c,v 1.1 1992/11/28 22:41:25 terry Exp $";
#endif


/*
 * X interface to Emacs.
 * Copyright (C) 1992  Terry Jones
 * (Based (heavily) on the taglist facility written by Brad Mears)
 *
 * This file is part of ebuttons
 *
 * Ebuttons is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 1, or (at your option)
 * any later version.
 *
 * The GNU General Public License can be obtained via anonymous ftp from
 * prep.ai.mit.edu as pub/gnu/COPYING or pub/gnu/COPYING-2.
 *
 * Ebuttons is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

/*
 * This provides a point-and-click command interface to an 
 * emacs session.  This won't be much use without ebuttons.el.
 * See the README for details on use.
 */

#include <stdio.h>

#include <X11/Intrinsic.h>
#include <X11/StringDefs.h>
#include <X11/Shell.h>

#ifdef HAVE_XAW3D
#include <X11/Xaw3d/Box.h>
#include <X11/Xaw3d/Command.h>
#else
#include <X11/Xaw/Box.h>
#include <X11/Xaw/Command.h>
#endif

static void create_buttons();
static void button_press();
void toggle_visibility();

typedef struct {
    String labels[MAX_BUTTONS];
    String commands[MAX_BUTTONS];
} AppResources;

static AppResources appResources;
static Widget toplevel;            

#include "ebuttons.h"

int 
main(argc, argv)
int argc;
char **argv;
{
   XtAppContext appContext;

   toplevel = XtAppInitialize(&appContext, "ebuttons", NULL, 0, &argc, argv, NULL, NULL, 0);
   XtGetApplicationResources(toplevel, (XtPointer)&appResources, resources, XtNumber(resources), NULL, 0);
   create_buttons(toplevel);
   XtAppAddInput(appContext, fileno(stdin), XtInputReadMask, toggle_visibility, NULL);
   XtRealizeWidget(toplevel);
   XtAppMainLoop(appContext);
   return 0;
}


/*
 * Callback for the configurable command buttons.  The quit button
 * has no associated command.
 */
static void 
button_press(w, command, dummy)
Widget  w;
XtPointer command; 
XtPointer dummy;
{
   if (!command){
      exit(0);
  }

   printf("%s", (String)command);
   fflush(stdout);
   return;
}


static void 
create_buttons(parent)
Widget parent;
{
    Widget cmds[MAX_BUTTONS];
    Widget cmdbox;              
    Widget quit;
    int i;
    Dimension width;
    Dimension max_width = 0;
    
    cmdbox = XtCreateManagedWidget("cmdbox", boxWidgetClass, parent, NULL, 0);
    
    for (i = 0; i < MAX_BUTTONS; i++) {
	if (appResources.labels[i] && appResources.commands[i]){
	    cmds[i] = XtCreateManagedWidget(appResources.labels[i], commandWidgetClass, cmdbox, NULL, 0);
	    XtAddCallback(cmds[i], XtNcallback, button_press, (XtPointer)appResources.commands[i]);
	
	    XtVaGetValues(cmds[i], XtNwidth, &width, NULL);
	    
	    if (max_width < width){
		max_width = width;
	    }
	}
    }
    
    quit = XtCreateManagedWidget("Quit", commandWidgetClass, cmdbox, NULL, 0);
    XtAddCallback(quit, XtNcallback, button_press, NULL);
    
    XtVaGetValues(quit, XtNwidth, &width, NULL);
    
    if (max_width < width){
	max_width = width;
    }
    
    /* Set all the buttons to the same width */
    for (i = 0; i < MAX_BUTTONS; i++) {
	if (appResources.labels[i] && appResources.commands[i]){
	    XtVaSetValues(cmds[i], XtNwidth, max_width, NULL);
	}
    }
    
    XtVaSetValues(quit, XtNwidth, max_width, NULL);
    
    return;
}


void 
toggle_visibility(client_data, dummy1, dummy2)
XtPointer client_data;
int *dummy1;
XtInputId *dummy2;
{
   char buf[BUFSIZ];
   static int visible = 1;

    /* Read input from stdin.  We know it is ready since we have been called. */
   
   if (!fgets(buf, BUFSIZ, stdin)){
       /* This goes to stdout, not stderr. */
       printf("could not read from stdin!\n");
       exit(0);
   }

   if (visible) {
       XtUnmapWidget(toplevel);
       visible = 0;
   }
   else {
       XtMapWidget(toplevel);
       visible = 1;
   }
   
   return;
}

